Helper/PaymentMethods.php

```diff
@@ -179,6 +187,29 @@ protected function fetchPaymentMethods(?string $country = null, ?string $shopper
         }
 
         $paymentMethods = $responseData['paymentMethods'];
+
+        $allowMultistoreTokens = $this->configHelper->getAllowMultistoreTokens($store->getId());
+
+        if (!$allowMultistoreTokens && isset($responseData['storedPaymentMethods'])) {
+            $customerId = $quote->getCustomerId();
+            $searchCriteria = $this->searchCriteriaBuilder
+                ->addFilter('customer_id', $customerId)
+                ->create();
+
+            $paymentTokens = $this->paymentTokenRepository->getList($searchCriteria)->getItems();
+
+            $gatewayTokens = array_map(function ($paymentToken) {
+                return $paymentToken->getGatewayToken();
+            }, $paymentTokens);
+
+            if (isset($responseData['storedPaymentMethods'])) {
+                $storedPaymentMethods = $responseData['storedPaymentMethods'];
+                $responseData['storedPaymentMethods'] = array_filter($storedPaymentMethods, function ($method) use ($gatewayTokens) {
+                    return in_array($method['id'], $gatewayTokens);
+                });
+            }
+        }
```

Comment:
What would you think about moving this filtering algorithm to a private method to increase the maintainability and readability?